[HUDI-3998] Fix getCommitsSinceLastCleaning failed when async cleaning

dongkelun opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contribute/how-to-contribute before opening a pull request.*

## What is the purpose of the pull request

Fix getCommitsSinceLastCleaning failed when async cleaning

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

nsivabalan commented:
good find on the bug. 
if my understanding is right, lastCompletedCommitTimestamp actually refers to last CompletedCommit w/o any holes. i.e. before the last completed commit, there should not any inflight commit. either its completed or rolledback. 

Also, can you confirm the changes are backwards compatible ?

dongkelun commented:
@nsivabalan  As the comments in the code,`lastCompletedCommitTimestamp`  is the last completed commit timestamp before clean. There is no guarantee that it is the last one on the whole timeline,because we can't ensure that there is no completed commit after clean,It is bound to clean.

 `before the last completed commit, there should not any inflight commit. either its completed or rolledback.`As for this, I think it is related to the rollback mechanism. As long as we ensure that we will rollback before each commit, this is true
 
For version compatibility, I verified it locally, it is backward compatible .Because we added a default value for the value of `lastCompletedCommitTimestamp`  in `HoodieCleanerPlan` and  `HoodieCleanMetadata`.In this way, 'lastCompletedCommitTimestamp' uses the default value("") when serializing files of older versions.When 'lastCompletedCommitTimestamp'is an empty string,
`numCommits = commitTimeline.findInstantsAfter(lastCompletedCommitTimestamp).countInstants();` returns the number of all completed commits,The subsequent clean operation can run normally

nsivabalan commented:
@dongkelun : sorry, I dropped the ball on this. can you rebase w/ latest master. I can review it and we can get it in for 0.12.1.

nsivabalan reviewed (commented):
LGTM. if you can rebase w/ latest master and once CI succeeds, we can land it in

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanActionExecutor.java:
```diff
@@ -64,11 +65,16 @@ private int getCommitsSinceLastCleaning() {
+        String lastCompletedCommitTimestamp = cleanMetadata.getLastCompletedCommitTimestamp();
+        numCommits = commitTimeline.findInstantsAfter(lastCompletedCommitTimestamp).countInstants();
+      } catch (IOException e) {
+        throw new HoodieIOException(e.getMessage(), e);
```
exception (2nd arg) will carry the msg anyways. can we fix the first argument w/ custom msg ("Parsing of last clean instant " + lastCleanInstant.get() + " failed")

dongkelun replied:
Good idea, Done

nsivabalan commented on hudi-common/src/main/avro/HoodieCleanMetadata.avsc:
```diff
@@ -23,6 +23,7 @@
      {"name": "timeTakenInMillis", "type": "long"},
      {"name": "totalFilesDeleted", "type": "int"},
      {"name": "earliestCommitToRetain", "type": "string"},
+     {"name": "lastCompletedCommitTimestamp", "type": "string", "default" : ""},
```
why empty string. we could go w/ null ?

dongkelun replied:
As I mentioned earlier, the main reason is to be compatible with the old version. The default value should not be null,
```scala
numCommits = commitTimeline.findInstantsAfter(lastCompletedCommitTimestamp).countInstants()
```
Null and empty return different results

nsivabalan commented on hudi-common/src/main/avro/HoodieCleanerPlan.avsc:
```diff
@@ -42,6 +42,11 @@
+    {
+      "name": "lastCompletedCommitTimestamp",
+      "type": "string",
+      "default" : ""
```
same here.

nsivabalan commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -502,14 +502,20 @@ public void handle(@NotNull Context context) throws Exception {
+            HoodieTimeline localTimeline =
+                viewManager.getFileSystemView(context.queryParam(RemoteHoodieTableFileSystemView.BASEPATH_PARAM)).getTimeline();
+            HoodieTimeline afterLastInstantTimeLine = localTimeline.findInstantsAfter(lastInstantTs).filterCompletedInstants();
+            if (!(afterLastInstantTimeLine.countInstants() == 1
```
So, are we making an exception to just 1 case here. 
i.e. when client has 1 extra commit compared to timeline server and if that is a clean action, we don't trigger a refresh ? 
can you move this to a separate method. I see we might potentially add more cases in here going forward.

dongkelun replied:
`syncIfLocalViewBehind` is triggered before this. This judgment is added here mainly to avoid throwing exceptions during asynchronous clean, otherwise the program cannot execute normally.

I will try to move it to a separate method

dongkelun replied:
Done

nsivabalan reviewed (commented):
A high level thought. I appreciate you parametrized most tests for both sync and async cleaning. But do keep in mind that, these are going to add to our test run-time. So, may be take another look to see if we need to make all cleaner tests run for both sync and async. Enable for those where it matters and for others probably running it for sync cleaning should suffice.

dongkelun commented:
> A high level thought. I appreciate you parametrized most tests for both sync and async cleaning. But do keep in mind that, these are going to add to our test run-time. So, may be take another look to see if we need to make all cleaner tests run for both sync and async. Enable for those where it matters and for others probably running it for sync cleaning should suffice.

Unnecessary asynchronous cleaning test cases have been removed

nsivabalan commented:
also, a good practice to follow. whenever you are addressing feedback, try to add it as new commits. Easier for reviewer to re-review just the new changes. if not, I have to review entire patch again. If not, I will click on newer commits and will review only the newly changed code.

nsivabalan commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -539,4 +543,19 @@ public void handle(@NotNull Context context) throws Exception {
+  /**
+   * Determine whether to throw an exception when local view of table's timeline is behind that of client's view.
+   */
+  private boolean shouldThrowExceptionIfLocalViewBehind(HoodieTimeline localTimeline, String lastInstantTs) {
```
shouldn't we call this from within isLocalViewBehind(). already we have timline there right. we don't need to re-instantiate the timeline again in L507.

dongkelun replied:
Sorry, if it is called in `isLocalViewBehind`, there is a timeline, but in the `handle` method, I don't see where there is a timeline, and the original code instantiates the timeline in `errMsg`

dongkelun replied:
My idea is to judge whether to throw an exception if `isLocalViewBehind` returns true. Because the `isLocalViewBehind` method is also used in the `syncIfLocalViewBehind` method, I am not sure whether it is appropriate to directly modify the logic of the `isLocalViewBehind`  method

dongkelun replied:
For example, when there is only one more .clean.completed, it should also be synchronized

dongkelun replied:
I modified the logic of `shouldThrowExceptionIfLocalViewBehind` method. I think it's better

nsivabalan reviewed (commented):
1 minor comment.

dongkelun commented:
> also, a good practice to follow. whenever you are addressing feedback, try to add it as new commits. Easier for reviewer to re-review just the new changes. if not, I have to review entire patch again. If not, I will click on newer commits and will review only the newly changed code.
OK, sorry, I thought it was OK to click compare after 'force pushed'. I mistakenly thought that 'force pushed' would look cleaner. I don't know you reviewed it by comparing two commits. I'll pay attention in the future
